src/provider/mod.rs

```diff
@@ -112,6 +116,150 @@ impl CustomGetHandler for () {
+/// calling `next`.
+///
+/// For collections that do allow efficient random access, the `skip` method can be
+/// used to move some internal offset.
+pub trait CollectionParser: Send + Default + 'static {
+    /// The iterator that allows to iterate over the collection.
+    type CollectionIter: CollectionIter;
+    /// Future returned by [`CollectionParser::parse`].
+    ///
+    /// For an initial implementation where you don't care that much about allocations,
+    /// this can be just `BoxFuture`. This would be the default if defaults were allowed.
+    type NewFuture: Future<Output = anyhow::Result<Self::CollectionIter>> + Send + 'static;
+    /// Parse a collection with this parser
+    fn parse<R: AsyncSliceReader + Send + 'static>(format: u64, reader: R) -> Self::NewFuture;
+}
+
+/// An iterator over a collection. Allows to get the next hash or skip a number of hashes.
+///
+/// This does not implement `Iterator` or `Stream` because of the extra `skip` method.
+///
+/// It uses state passing style.
+pub trait CollectionIter: Sized + Send + 'static {
+    /// Future returned by [`CollectionIter::next`].
+    ///
+    /// For an initial implementation where you don't care that much about allocations,
+    /// this can be just `BoxFuture`. This would be the default if defaults were allowed.
+    type NextFuture: Future<Output = (Self, anyhow::Result<Option<Hash>>)> + Send + 'static;
+    /// Future returned by [`CollectionIter::skip`].
+    ///
+    /// For an initial implementation where you don't care that much about allocations,
+    /// this can be just `BoxFuture`. This would be the default if defaults were allowed.
+    type SkipFuture: Future<Output = (Self, anyhow::Result<()>)> + Send + 'static;
+    /// Get the next hash in the collection.
+    fn next(self) -> Self::NextFuture;
+    /// Skip a number of hashes in the collection.
+    fn skip(self, n: u64) -> Self::SkipFuture;
+}
+
+impl CollectionIter for Never {
+    type NextFuture = future::Ready<(Self, anyhow::Result<Option<Hash>>)>;
+    type SkipFuture = future::Ready<(Self, anyhow::Result<()>)>;
+    fn next(self) -> Self::NextFuture {
+        unreachable!()
+    }
+    fn skip(self, _n: u64) -> Self::SkipFuture {
+        unreachable!()
+    }
+}
+
+/// Define CustomCollection for (). This is for when you don't want to use support collections at all.
+impl CollectionParser for () {
+    type CollectionIter = Never;
+
+    type NewFuture = future::Ready<anyhow::Result<Never>>;
+
+    fn parse<R: AsyncSliceReader + Send + 'static>(_format: u64, _reader: R) -> Self::NewFuture {
+        future::err(anyhow::anyhow!("collections not supported"))
+    }
+}
+
+/// Parser for the current iroh default collections
+#[derive(Debug, Clone, Copy, Default)]
+pub struct DefaultCollectionParser;
+
+/// Iterator for the current iroh default collections
+#[derive(Debug, Clone)]
+pub struct ArrayCollectionIter {
+    hashes: Box<[Hash]>,
+    offset: usize,
+}
```

Comment:
If we really want to handle any size collection then this wouldn't hold an array of hashes but would rather be a `&[Hash]` and it could be referring to a mmap or so for actual data.  Is it possible to make this or are the lifetimes just not doable?

Reply:
so you don't feel like making `IrohCollectionParser` deal with arbitrarily large collections out of the box?  I'm also fine with creating a follow up issue if the answer is "yes, but rather not in this PR"

Reply:
The answer is "yes, but rather not in this PR" indeed...